posts/exploring-iot-infrastructure-support-in-aws-part2.md

```diff
@@ -0,0 +1,450 @@
+}
+```
+
+Getting device state of the selected device:
+
+```Java
+@Override
+protected AsyncTaskResult<String> doInBackground(Void... voids) {
+    // Setting up our request for shadow state
+    GetThingShadowRequest getThingShadowRequest = new GetThingShadowRequest();
+    getThingShadowRequest.setThingName(name);
+
+    // Getting our shadow and parsing into string
+    GetThingShadowResult getThingShadowResult = dataClient.getThingShadow(getThingShadowRequest);
+    byte[] bytes = new byte[getThingShadowResult.getPayload().remaining()];
+    getThingShadowResult.getPayload().get(bytes);
+    result = new AsyncTaskResult<>(new String(bytes));
+    return result;
+}
+
+@Override
+protected void onPostExecute(AsyncTaskResult<String> result) {
+    // Updating our ThingDetails UI with Shadow Information
+    JSONObject parsedJson = new JSONObject(result.getResult());
+    String parsedToString = parsedJson.toString(spacesToIndentEachLevel);
+    ((ThingDetailsActivity) context).updateJson(parsedToString);
+}
+```
+## Android activity displaying Shadow state for an IoT device
+
+![Android Screenshot](https://raw.githubusercontent.com/BenEdridge/blog-usa/master/images/2018/07/aws_iot_app_screenshot_shadow.png)
+
+You can can see we are getting the the attributes associated with the device including the id, mac, temperature and humidity. This is retrieved from the device shadow state.
+
+# Show me the Code
+
+- [MongooseOS Project on ESP](https://github.com/BenEdridge/Ippon_IoT_ESP) with the required Terraform files and docs.
+- [Android application](https://github.com/BenEdridge/Ippon_IoT_Android) for interacting with our AWS IoT devices once the initial ESP project is setup.
+
+## Potential extensions or modifications of the Applications:
+
+- Custom alerts when changes are occur Eg. If temperature > 30 send email/SMS
+- Toggle device settings from the mobile app
+- Set up more devices grouping them and allowing management of groups through the app
+- Setup a Greengrass core and use this manage the devices instead
+- Experiment with ML inference using results or pass them into Lambda function
+
+# What next?
+
+ AWS has some solid IoT infrastructure but also lacks in certain areas around initial setup and device documentation. There is complexity in setup especially with Greengrass and AWS FreeRTOS. Ideally this will be streamlined in the future creating an easier setup with new ways to set up TLS mutual authentication and perhaps some other authentication methods if needed.
+ 
+ There is also massive hole in CloudFormation and deployment options for Greengrass. AWS IoT Core isn't so bad but still lacks a huge amount of features. Hopefully in the future we see more support by AWS CloudFormation and Terraform. Right now any complex setup relies on the CLI, AWS Console or one of the supported SDK libraries. This means you need to setup your own way of managing stack state and deployment.
```

Comment:
I'd change this to something like `Cloudformation deployment options for Greengrass are particularly limited.`